smartconfig/configuration.py

```diff
@@ -0,0 +1,173 @@
+    and will also be overwrote in future by newly loaded configurations.
+
+    The path used by an entry is by default the `__name__` attribute of the module it is defined in.
+
+    Attributes:
+        _path_override:
+            Can be used to override the default `_path` setting.
+        _path:
+            Path used to reference this entry in the configuration files. Shouldn't be manually set.
+        _configuration_mapping:
+            Current mapping of attributes. Shouldn't be manually set.
+        _defined_entries:
+            List of attributes being registered on the class,
+            either by giving a concrete value or a typehint. Shouldn't be manually set.
+    """
+
+    _path_override: Optional[str] = None
+
+    _path: str
+    _configuration_mapping: _EntryMapping
+    _defined_entries: List[str]
+
+    @classmethod
+    def _register_entry(cls):
+        """Set the `_path` attribute and register the entry."""
+        cls._path = cls._path_override or inspect.getmodule(cls).__name__
+        if cls._path in _register.configuration_for_module:
+            raise DuplicateConfiguration(f"The entry {cls._path} already exist.")  # TODO: Add an FAQ link.
+
+        _register.configuration_for_module[cls._path] = cls
+
+    @classmethod
+    def _fetch_configuration(cls):
+        """Load the configuration from the register and apply it."""
+        if cls._path in _register.global_configuration:
+            cls._patch_entries(_register.global_configuration[cls._path])
+
+    @classmethod
+    def _check_undefined_entries(cls):
+        """Raise `ConfigurationKeyError` if any attribute doesn't have a concrete value."""
+        for attribute in cls._defined_entries:
+            if attribute not in cls._configuration_mapping:
+                raise ConfigurationKeyError(f"Entry {attribute} isn't defined.")
+
+    @classmethod
+    def _patch_entries(cls, patch: _EntryMapping):
+        """
+        Override attributes using the given patch.
+
+        Args:
+            patch: Mapping of attribute names to their values
+
+        Raises:
+            ConfigurationKeyError: Trying to override an undefined attribute.
+        """
+        for key, value in patch.items():
+            if key not in cls._defined_entries:
+                raise ConfigurationKeyError(f"Entry {cls._path} doesn't define any {key} entry.")
+
+            cls._configuration_mapping[key] = value
+
+    def __init_subclass__(cls, **kwargs):
+        """
+        Initialize the new entry.
+
+        Raises:
+            DuplicateConfiguration: An entry is already registered for this path, use `_path_override`.
+            ConfigurationKeyError: An attribute doesn't have a concrete value.
+        """
+        super().__init_subclass__(**kwargs)
+
+        cls._register_entry()
+        cls._fetch_configuration()
+        cls._check_undefined_entries()
+
+    def __init__(self):
```

Comment:
This function will never return anything, since it will always raise upon usage of this dunder.
```suggestion
    def __init__(self) -> NoReturn:
```